File: clientA/src/app/core/services/thema.service.ts
import { HttpClient, HttpParams } from '@angular/common/http';
import { inject, Service } from '@angular/core';
import { PageModel } from '../models/page.model';
import { ThemaModel } from '../models/thema.model';

@Service()
export class ThemaService {
  private readonly url = '/api/themas';
  private readonly http = inject(HttpClient);

  getPage(page: number, size: number) {
    const params = new HttpParams().set('page', page).set('size', size);
    return this.http.get<PageModel<ThemaModel>>(this.url, { params });
  }

  getById(id: number) {
    return this.http.get<ThemaModel>(`${this.url}/${id}`);
  }

  create(thema: ThemaModel) {
    return this.http.post<ThemaModel>(this.url, { name: thema.name });
  }

  update(thema: ThemaModel) {
    return this.http.put<ThemaModel>(this.url, thema);
  }

  delete(id: number) {
    return this.http.delete<void>(`${this.url}/${id}`);
  }
}

File: LanguageCard/src/main/java/com/maryna/LanguageCard/Services/ThemaService.java
package com.maryna.LanguageCard.Services;

import com.maryna.LanguageCard.Models.PageModel;
import com.maryna.LanguageCard.Models.PageRequestModel;
import com.maryna.LanguageCard.Models.ThemaModel;
import com.maryna.LanguageCard.Repositories.ThemaCardRepository;
import com.maryna.LanguageCard.Repositories.ThemaRepository;
import org.apache.coyote.BadRequestException;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
public class ThemaService {
    private final ThemaRepository _themaRepository;
    private final ThemaCardRepository _themaCardRepository;

    public ThemaService(ThemaRepository themaRepository, ThemaCardRepository themaCardRepository) {
        _themaRepository = themaRepository;
        _themaCardRepository = themaCardRepository;
    }

    public PageModel<ThemaModel> getPage(PageRequestModel pageRequest) {
        var items = _themaRepository.getPage(pageRequest.limit(), pageRequest.offset());
        var total = _themaRepository.count();
        return new PageModel<>(items, pageRequest.page(), pageRequest.size(), total);
    }

    public ThemaModel getById(int id) throws BadRequestException {
        return _themaRepository.getById(id)
                .orElseThrow(() -> new BadRequestException("There is no such a theme!"));
    }

    public ThemaModel create(String name) {
        return _themaRepository.create(name);
    }

    @Transactional
    public ThemaModel update(ThemaModel themaModel) throws BadRequestException {
        if (!_themaRepository.exists(themaModel.getId())) {
            throw new BadRequestException("There is no such a theme!");
        }
        _themaRepository.update(themaModel);
        return getById(themaModel.getId());
    }

    @Transactional
    public void delete(int id) throws BadRequestException {
        if (!_themaRepository.exists(id)) {
            throw new BadRequestException("There is no such a theme!");
        }
        if (_themaCardRepository.hasCards(id)) {
            throw new BadRequestException("This theme has cards!");
        }
        _themaRepository.delete(id);
    }
}
